### ISSUE TITLE
Video and Voice Calling

### ISSUE THREAD

**Issue by viasux** [ajbura/cinny#528]
### Describe the problem

Currently, there is no way to voice call someone in Cinny.

### Describe the solution you'd like

An option to voice call ppl :)

### Alternatives considered

Using another matrix client

**Comment by menturion** [comment]
+1 

Support/integration of Element Call (https://github.com/vector-im/element-call).

**Comment by C0ffeeCode** [comment]
https://matrix.to/#/!rgQbWTfsoUcbvEDAvs:matrix.org/$IxJSHKXwmcSkfKgrybxlJb6hC-THyPD4oP3Ew-_dlyo?via=matrix.org&via=envs.net&via=kde.org

> There are no plans for calls right now. But it's probably going to be the new matrix inhouse VoIP calls and not jitsi.

**Comment by reesericci** [comment]
This would be great, as I could finally replace element with something much nicer to use IMO. I currently have to use element to talk to my friends over VoIP.

**Comment by ilyakooo0** [comment]
I want to implement this. Does anyone have any pointers? (I have no experience with matrix APIs).

**Comment by kfiven** [comment]
This is blocked by https://github.com/matrix-org/matrix-js-sdk/pull/2553

**Comment by FarisZR** [comment]
the pr is now merged

**Comment by viasux** [comment]
> the pr is now merged

Could you link it?

**Comment by FarisZR** [comment]
https://github.com/matrix-org/matrix-js-sdk/pull/2553

**Comment by reesericci** [comment]
@ilyakooo0

**Comment by ilyakooo0** [comment]
I am currently preoccupied by other projects, so I will not be implementing this in the foreseeable future. 

I will post a comment if I decide to pick this up again.

**Comment by reesericci** [comment]
no worries!

**Comment by viasux** [comment]
> +1
> 
> Support/integration of Element Call (https://github.com/vector-im/element-call).

I'd like to mention that I think using element call as a backend would be great, but it would make more sense to develop a user interface for it which follows the cinny design language. This could always be added down the line after just sloppily importing it though, so I'm not too worried about it, but one of the biggest strengths of cinny is how uniform its design language is.

**Comment by C0ffeeCode** [comment]
Unpopular opinion:
Call functionality might bloat Cinny

**Comment by SteveGremory** [comment]
has any progress been made on this topic?

**Comment by viasux** [comment]
> has any progress been made on this topic?

Pretty sure none has

**Comment by SteveGremory** [comment]
well that's kinda sad

**Comment by viasux** [comment]
> well that's kinda sad

Still the one thing holding me back from really daily driving matrix

(including screen sharing with audio, but id at least use it a lot more if any form of calling was implemented)

**Comment by IceDBorn** [comment]
https://github.com/cinnyapp/cinny/pull/1163

**Comment by FarisZR** [comment]
The problem is matrix itself is moving away from jitsi and into its own [matrix VOIP](https://element.io/blog/introducing-native-matrix-voip-with-element-call/)

**Comment by IceDBorn** [comment]
@FarisZR This is a temporary solution until Cinny devs figure out how they're going to handle calls, I do not want my draft to be merged.

**Comment by K4LCIFER** [comment]
Would adding in voice/video call functionality to Cinny include Discord-like voice/video rooms ("Rooms" in the sense of being able to pop in and out)?

**Comment by rwjack** [comment]
So should we wait for Element Call, hoping that it will replace both Jitsi for conference calls, and Coturn servers for 1:1 calling, or is Element Call only replacing Jitsi?

If that's the case, we'd still be missing the regular 1:1 call feature in Cinny, using self-hosted Coturn servers, as available in Element.

**Comment by FarisZR** [comment]
AFAIK element call is just element 1:1 calls but for groups.

**Comment by Cyrus-Harding** [comment]
> Would adding in voice/video call functionality to Cinny include Discord-like voice/video rooms ("Rooms" in the sense of being able to pop in and out)?

Element already has this functionality.

**Comment by rwjack** [comment]
> Element already has this functionality.

He was probably referring to the UI simplicity of Discord, as in showing small Avatars that are in the voice channel call, and the ability to join by clicking the Channel, and leave with a disconnect button on the bottom of the UI.

**Comment by K4LCIFER** [comment]
> > Element already has this functionality.
> 
> He was probably referring to the UI simplicity of Discord, as in showing small Avatars that are in the voice channel call, and the ability to join by clicking the Channel, and leave with a disconnect button on the bottom of the UI.

Correct, this is what I am referring to. Popping in and out of a persistent channel is an incredibly useful feature.

**Comment by 42willow** [comment]
Element has video rooms already and they're pretty stable as far as I know. Is there anything blocking this from being implemented?

EDIT: Never mind, element uses Jitsi for video rooms (for now until element call is stable)

<del>I'd be happy to work on this if someone roughly pointed me in the right direction, I have quite a bit of spare time at the moment.</del>
My three weeks of spare time are over but I can still help

**Comment by FarisZR** [comment]
I think you should just work on embedding element Call inside Cinny similar to how Jitsi currently works in element.
This should be quick and easy, and Element call is almost ready anyway.
You can go further and try to implement it 'natively' but I'm not sure if this is possible or not.

In any case, the project should definitely skip implementing Jitsi.

**Comment by 42Willow** [comment]
> I think you should just work on embedding element Call inside Cinny similar to how Jitsi currently works in element. This should be quick and easy, and Element call is almost ready anyway. You can go further and try to implement it 'naively' but I'm not sure if this is possible or not.
> 
> In any case, the project should definitely skip implementing Jitsi.

Element Call didn't work for me when I tried it last week :thinking:. I probably won't be able to do this all by myself as I'm not that adept at programming. Maybe we could setup a fork and a matrix room?

**Comment by SnoutBug** [comment]
I am also in favor of adding VoIP to Cinny, however I am not sure whether implementing Element Call is the right way to go. Currently Element Call is a pain to set up and I would not consider version 0.5 to be "almost ready".

Element Call feels unfinished and more like a prove of concept, rather than a feature which is ready to be implemented. If I am not mistaken, implementing Element Call in Cinny would mean to add a Room which hosts an iframe to `call.element.io` or any other URL the hoster specified. In my mind this would be no different to adding Jitsi to Cinny, but correct me if I'm wrong about that.

In that case anyone wanting a call feature could just host Element Call separately and link it to their Matrix instance instead. While this is probably not the prettiest way of doing things, its probably more pretty than including Element Call as an iframe in Cinny which breaks the theme consistency and having to implement the feature anew once a proper solution arrives, eg. native VoIP via Matrix.

**Comment by fbievan** [comment]
I think we should have 1 on 1 voice calling until group calling is fully fledged out.

**Comment by deepadmax** [comment]
The proposals will probably be finished in a couple months and it would be better to wait than to implement a soon-to-be-dated version of calls.

I think it would be a great disservice to the much appealing design of Cinny to simply embed Element Call rather than to follow it to its conclusion and implement video rooms like Discord voice channels.

**Comment by Beryesa** [comment]
In the meantime an Element call widget would be really useful though.
Just to scale, Element call is progressing really well already.

**Comment by 42willow** [comment]
> The proposals will probably be finished in a couple months and it would be better to wait than to implement a soon-to-be-dated version of calls.
> 
> I think it would be a great disservice to the much appealing design of Cinny to simply embed Element Call rather than to follow it to its conclusion and implement video rooms like Discord voice channels.

@deepadmax where can these proposals be found? Apparently element call is production ready now :tada: https://element.io/blog/we-have-lift-off-element-x-call-and-server-suite-are-ready/
I've tested it at [call.element.io](https://call.element.io/) and it seems to be running smoothly without any noticeable issues.

Regarding the idea of implementing Discord-like voice channels: Has there been any discussion about this with the core maintainers of Cinny? It would be good to agree on this first, so that if someone starts working on it, they won’t have to make significant changes later on. Personally I think it's a great idea, but I'm not yet sure if Element Call is fast enough to make popping in and out of a call as seamless as it is on Discord. Another thing to note is on Discord, the UI is updated with your profile appearing in the call list *before* you have actually connected to the call backend.

**Comment by SnoutBug** [comment]
I am not sure what the blog post is talking about. Maybe I am misunderstanding the procedure, but the proposals for VoIP via Matrix still have not been merged.
Element Call has worked before, but I could not find any significant change. So far it is still required to host a separate endpoint for calls (develop.element.io requires call.element.dev for calls)
Also it seems Element Call is still in early development (v0.6.6). element-x-ios is `[...] work in progress. The app does not cover yet all functionalities we expect.` element-x-android `[...] is currently in a pre-alpha release stage with only basic functionalities.` 

I have come to the conclusion that I don't understand what I should expect from Element-X, Element Call or VoIP in Matrix.

**Comment by Destinyg133** [comment]
This would be great addition to cinny

**Comment by nac0n** [comment]
Would love to see it happen in Cinny.
I fell in love with the UI direction and want to use it as my standard Matrix client. "No voice" is holding me back at the moment unfortunately.

**Comment by omiinaya** [comment]
+1 here. Would love to see voice and video on Cinny

**Comment by ghost** [comment]
Cannot spot Voice Calls in the public roadmap... Any news on this? Thanks!

**Comment by SnoutBug** [comment]
> Cannot spot Voice Calls in the public roadmap... Any news on this? Thanks!

Last time I checked Element Call ist still very far from being finished. I wouln't blame anyone not wanting to implement this yet.

**Comment by ghost** [comment]
> > Cannot spot Voice Calls in the public roadmap... Any news on this? Thanks!
> 
> Last time I checked Element Call ist still very far from being finished. I wouln't blame anyone not wanting to implement this yet.

Of course I'm not "blaming" anyone, why would I? :)
Element Call seems going ahead quite well: https://github.com/element-hq/element-call/releases/

**Comment by alexander-potemkin** [comment]
> > > Cannot spot Voice Calls in the public roadmap... Any news on this? Thanks!
> > 
> > 
> > Last time I checked Element Call ist still very far from being finished. I wouln't blame anyone not wanting to implement this yet.
> 
> Of course I'm not "blaming" anyone, why would I? :) Element Call seems going ahead quite well: https://github.com/element-hq/element-call/releases/

To add on that: Element Call is now an official way for calling in many scenarios, as Element 'Classic' is about to deprecated and all of the development efforts are now at Element X - which, from what I can tell, only support Element Call.

Happy to be corrected!

**Comment by 32bitx64bit** [comment]
Any news? Element call seems to be more fleshed out and more stable than it  was 2 years ago.
Would love to see it get implemented.

**Comment by OTonGitHub** [comment]
Element calls work really well now https://github.com/element-hq/element-call/releases/ The only thing holding back Cinny right now over other clients is the ability to do Calls. The UI is so flushed out  and easier for new users to pick up, so much better than any other desktop clients. And being compatible with the new Element X's protocol would make cross-platform calls possible too.. Can't wait for this to get implemented. Moreover, is there an update announcement channel available for Cinny on Matrix perhaps?

**Comment by kfiven** [comment]
Re: To folks commenting on PR related to this feature.

This is on our todo list for upcoming features.
Initial thoughts are:
1. We need to make designs for this feature so that it aligns with rest of the UI/UX.
2. Need to figure out if we should embed the element call instance provided by Element or host our own.
    a. If use element hosted, should we talk to element first as we will be putting burden on their resources?
    b. If hosting our own, how should we go that we are not hosting a free call service that runs out our pocket (because it is going to use some resources)

**Comment by Destinyg133** [comment]
> Re: To folks commenting on PR related to this feature.
> 
> This is on our todo list for upcoming features. Initial thoughts are:
> 
>     1. We need to make designs for this feature so that it aligns with rest of the UI/UX.
> 
>     2. Need to figure out if we should embed the element call instance provided by Element or host our own.
>        a. If use element hosted, should we talk to element first as we will be putting burden on their resources?
>        b. If hosting our own, how should we go that we are not hosting a free call service that runs out our pocket (because it is going to use some resources)

Why not do the same thing as Element and make it possible that users run their own EC
Nontheless, its great that it will have support from cinny

**Comment by subjugum** [comment]
Great work!

2. Let the users HS worry about the MatrixRTC stack, but perhaps display a warning if none is advertised in the well-known. Element closed off their public instance, too.

**Comment by 32bitx64bit** [comment]
👀

**Comment by shootie22** [comment]
> Great work!
> 
> 2. Let the users HS worry about the MatrixRTC stack, but perhaps display a warning if none is advertised in the well-known. Element closed off their public instance, too.

I also think this should be the way. In addition, I would add a very simple and straight forward explanation of how to host EC, perhaps simpler than on the official docs if possible, so that users don't have to dig through. 
  
Maybe also add a list of public EC instances, if that's a thing? That way the user could easily choose one without having to go straight to self hosting. Not sure if it's a thing though.

**Comment by alexander-potemkin** [comment]
I would absolutely need an ability to switch to a self-hosted instance - something I can do now with Element X via .well-known configuration on the server.

As for using Element Call of Element/Matrix... It looks like they push it - given you will be using they widget, you would just inherit they defaults ;)

**Comment by alexander-potemkin** [comment]
> Re: To folks commenting on PR related to this feature.
> 
> This is on our todo list for upcoming features. Initial thoughts are:

Thank you! Is there any possible timelines?

I enjoy using Cinny - it is the cleanest and most easy to use messenger. Calls and mobile support (https://github.com/cinnyapp/cinny/issues/2400) would drive it to a completely different level.

For the EC server - if that would be the only thing that stops, I would be glad to try to provide a ready to use server for Cinny.

**Comment by olivierlambert** [comment]
I agree, best client for chatting right now, calls are the only piece missing to be better than Element IMHO!

**Comment by MatthieuCoder** [comment]
Given that Element Call (the *only* implementation of MatrixRTC?) is made to be embeded in [apps using widgets](https://github.com/element-hq/element-call/tree/livekit?tab=readme-ov-file#-element-call-packaging) and given that editing the style of element call (which is using React, which usually mangles the css classes names depending on the bundler config) would be a burden, I don't really see Cinny changing the design of Element Call to match

**Comment by LinuxinaBit** [comment]
> Need to figure out if we should embed the element call instance provided by Element or host our own.

My impression from https://github.com/element-hq/element-call/blob/livekit/README.md is that Element calls are supposed to be, in the future, associated with the Matrix homeserver.
I really think you should use Element’s Element call instance for now, then support the homeserver selected one once that functionality becomes available.

Also because Element already uses Element Call afaik, I don’t think there’s any need to worry about infrastructure costs on Matrix’s end since they already expect Matrix users to use it.

**Comment by dan0v** [comment]
Here is [a very straightforward guide](https://willlewis.co.uk/blog/posts/deploy-element-call-backend-with-synapse-and-docker-compose/?ref=element.io) to adding Element Call support to a Matrix Synapse Homeserver. When running an Element Call instance beside the Homeserver, as seems to be the goal for Element, the following block of a modern Matrix Homeserver's `.well-known/matrix/client` file is there to inform clients such as Cinny of where to find the Element Call service associated with this Homeserver.:
```
"org.matrix.msc4143.rtc_foci":[
      {
         "type":"livekit",
         "livekit_service_url":"https://whatever.yourdomain.com"
      }
   ]
```

I am running Synapse myself following this guide and Element Call works perfectly with the Element web, desktop, and Element X android clients. I don't see why @GigiaJ 's implementation couldn't set the `elementCallUrl` (hard coded in `config.json` in the current PR) to this value from the Homeserver info (for example, [here](https://github.com/cinnyapp/cinny/blob/dev/src/app/cs-api.ts#L89)), leaving only merge conflicts to be resolved before this feature could be in users' hands. Extra custom styling on top of an embedded widget seems to me, to be extraneous to actually having functioning call support. As for many others, this is the last thing missing from Cinny for my switch from Element.

* Edit * - I have tested running Gigiaj's branch locally and it works for voice/video rooms, but there is no button to start a call with a user explicitly, or to start a call in a non-call room. Also, when being called by other users, an "Unsupported message" error is shown in the chat, rather than any sort of ringtone. My suggestion to add the element call URL to the config.json section is actually not required either, since the bundled element-call widget takes care of establishing the call correctly (the config.json is to override the widget rendering with a remote call frontend). It seems that the ability to initiate and receive calls is all that's missing, but the calling functionality within dedicated call channels works.

**Comment by morleyweb** [comment]
Cinny is honestly the best client I've seen so far.
Really only missing out on calling, which is probably holding it back from wider use.

**Comment by orodes** [comment]
is there an ETA on when this will be done?

**Comment by YoJames2019** [comment]
How is the progress on this? cinny is by far the best client i've used, the only thing holding it back is the lack of voice call + screen sharing support

**Comment by shootie22** [comment]
Adding some screenshots of the current state of [https://github.com/cinnyapp/cinny/pull/2335](https://github.com/cinnyapp/cinny/pull/2335)'s implementation of Element Call in Cinny:  

<img width="1613" height="885" alt="Image" src="https://github.com/user-attachments/assets/4a6b4999-b005-42e9-9ac2-6625ca0be1f5" />

<img width="1605" height="884" alt="Image" src="https://github.com/user-attachments/assets/19630811-ebd6-4118-a1c3-a710cee6b7f2" />

**Comment by Sollace** [comment]
> Adding some screenshots of the current state of [#2335](https://github.com/cinnyapp/cinny/pull/2335)'s implementation of Element Call in Cinny:

We have been running this for a couple months now and can confirm it works pretty well.

The only issue is the bad audio quality of Element's screenshare, but that is not the fault of cinny.

**Comment by alexander-potemkin** [comment]
> > Adding some screenshots of the current state of [#2335](https://github.com/cinnyapp/cinny/pull/2335)'s implementation of Element Call in Cinny:
> 
> We have been running this for a couple months now and can confirm it works pretty well.
> 
> The only issue is the bad audio quality of Element's screenshare, but that is not the fault of cinny.

That is fantastic! Any ETA for the Cinny official release, that will include it?

**Comment by Mondrethos** [comment]
Also really excited for this.

**Comment by Sollace** [comment]
> > > Adding some screenshots of the current state of [#2335](https://github.com/cinnyapp/cinny/pull/2335)'s implementation of Element Call in Cinny:
> > 
> > 
> > We have been running this for a couple months now and can confirm it works pretty well.
> > The only issue is the bad audio quality of Element's screenshare, but that is not the fault of cinny.
> 
> That is fantastic! Any ETA for the Cinny official release, that will include it?

Nope, sorry. I should clarify that I'm not at all associated with cinny.

**Comment by alexander-potemkin** [comment]
> Nope, sorry. I should clarify that I'm not at all associated with cinny.

Sure and thank you, @Sollace !

**Comment by hazre** [comment]
Not directly relevant but Discord will [require age verification globally next month](https://discord.com/press-releases/discord-launches-teen-by-default-settings-globally), and I think there is a likely hood of a mass exodus to other platforms. I personally think Element clients are a lost cause in many aspects in terms of UX (because it's trying to be many things at once) compared to what Discord users are used to. So other Matrix clients like Cinny might just be a decent replacement if blockers like voice calling and screen sharing are resolved some time soon.

**Comment by ZeroZeroOne-dev** [comment]
> Not directly relevant but Discord will [require age verification globally next month](https://discord.com/press-releases/discord-launches-teen-by-default-settings-globally), and I think there is a likely hood of a mass exodus to other platforms. I personally think Element clients are a lost cause in many aspects in terms of UX (because it's trying to be many things at once) compared to what Discord users are used to. So other Matrix clients like Cinny might just be a decent replacement if blockers like voice calling and screen sharing are resolved some time soon.

That's literally why I am here, I'm looking for an alternative, the reason Im passing element is bc you cant have an audio-only channels? Only video channels where by default cams are on...

**Comment by Mondrethos** [comment]
I am aware of several big communities that are moving from Discord including mine. Cinny is the most talked about client.



> > Not directly relevant but Discord will [require age verification globally next month](https://discord.com/press-releases/discord-launches-teen-by-default-settings-globally), and I think there is a likely hood of a mass exodus to other platforms. I personally think Element clients are a lost cause in many aspects in terms of UX (because it's trying to be many things at once) compared to what Discord users are used to. So other Matrix clients like Cinny might just be a decent replacement if blockers like voice calling and screen sharing are resolved some time soon.
> 
> That's literally why I am here, I'm looking for an alternative, the reason Im passing element is bc you cant have an audio-only channels? Only video channels where by default cams are on...

Does that Cinny PR let you do audio only rooms?

**Comment by YoJames2019** [comment]
> I am aware of several big communities that are moving from Discord including mine. Cinny is the most talked about client.
> 
> 
> 
> > > Not directly relevant but Discord will [require age verification globally next month](https://discord.com/press-releases/discord-launches-teen-by-default-settings-globally), and I think there is a likely hood of a mass exodus to other platforms. I personally think Element clients are a lost cause in many aspects in terms of UX (because it's trying to be many things at once) compared to what Discord users are used to. So other Matrix clients like Cinny might just be a decent replacement if blockers like voice calling and screen sharing are resolved some time soon.
> > 
> > That's literally why I am here, I'm looking for an alternative, the reason Im passing element is bc you cant have an audio-only channels? Only video channels where by default cams are on...
> 
> Does that Cinny PR let you do audio only rooms?


currently the default for audio rooms in that pr is audio only with the option of video and screensharing once youre in call, but the ui is a bit jank. I'm currently in the process of reworking it. about 90% done with that atm

**Comment by Sollace** [comment]
> I am aware of several big communities that are moving from Discord including mine. Cinny is the most talked about client.
> 
> > > Not directly relevant but Discord will [require age verification globally next month](https://discord.com/press-releases/discord-launches-teen-by-default-settings-globally), and I think there is a likely hood of a mass exodus to other platforms. I personally think Element clients are a lost cause in many aspects in terms of UX (because it's trying to be many things at once) compared to what Discord users are used to. So other Matrix clients like Cinny might just be a decent replacement if blockers like voice calling and screen sharing are resolved some time soon.
> > 
> > 
> > That's literally why I am here, I'm looking for an alternative, the reason Im passing element is bc you cant have an audio-only channels? Only video channels where by default cams are on...
> 
> Does that Cinny PR let you do audio only rooms?

My experience is you can turn off video before joining. But yeah, a UI that is audio-first with the option to screenshare (with audio that doesn't duck!) in the style of Discord would be perfect.


(duck meaning it decreases in volume/bitrate whenever someone else talks, not duck as in "this thing so ducking ducks")

**Comment by lizard-demon** [comment]
Just throwing in my 2 cents here. I'm here for discord migration.

**Comment by Lyranile** [comment]
Would love to see voice calling!

**Comment by HeadClot** [comment]
With discord shitting the bed. Expect allot of people to pop in and chime in on this topic. Just saying.

**Comment by throwing-shade** [comment]
Yeah! I'm very much tired of Discord's nonsense and would love to find some kind of small place for me and my friends. A voice call implementation would mean the world to us : )

**Comment by RayZ3R0** [comment]
Would be amazing if Cinny gets the vc feature soon, a lot of discord refugees would rush in. Currently using Cinny for chatting and Element for VC sadly. Cinny is simply a way better app for just chat

**Comment by shioprince** [comment]
I would also like voice/video rooms too.

**Comment by Maoilioss** [comment]
come one cinny, its ur time to shine! pplllss

**Comment by knightingalelmao** [comment]
this would be the single best github push when complete. can't wait to ditch discord for cinny

**Comment by aedelor** [comment]
Cinny really needs voice and video. So far it has been my favourite matrix client but my group uses more than just text chat.

**Comment by willh34** [comment]
I agree I'm trying to move from discord and this is the only thing stopping me
